swift/swift.py

```diff
@@ -52,102 +61,131 @@ def __init__(self, setup_env: dict, parent=None):
+        self.mainWindow.show()
 
-    def _init_bus(self, setup_bus: dict):
-        """Initializes global buses using set-up environment.
+    def load(self, appInfos: dict, busInfos: dict):
+        """Initializes swift system and loads the apps and buses.
         
         Args:
-            setup_bus: Set-up environment about bus.
+            appInfos: See "app" part of setupEnv in self.__init__().
+            busInfos: See "bus" part of setupEnv in self.__init__().
         """
-        for name, info in setup_bus.items():
-            # create a bus
-            if "timeout" in info:
-                timeout = info["timeout"]
-                bus = Bus(name, timeout)
-            else:
-                bus = Bus(name)
-            # set a slot of received signal to router
-            bus.received.connect(self._route_to_app)
-            bus.start()
-            # store the bus
-            self._buses[name] = bus
-            self._subscribers[name] = []
-
-    def _init_app(self, setup_app: dict):
-        """Initializes apps using set-up environment.
+        for name, info in busInfos.items():
+            self.createBus(name, info)
+        for name, info in appInfos.items():
+            self.createApp(name, info)
+
+    def createBus(self, name: str, info: dict):
+        """Creates a global bus using set-up environment.
         
         Args:
-            setup_app: Set-up environment about app.
+            name: A name of the bus.
+            info: A dictionary containing bus info. Each element is like below:
+              timeout: Desired timeout for blocking queue reading, in seconds.
+                If you want to set as default, set None.
         """
-        for name, info in setup_app.items():
-            # import the app module
-            path = info.get("path", ".")
-            mod_name, cls_name = info["module"], info["class"]
-            with _add_to_path(os.path.dirname(path)):
-                module = importlib.import_module(mod_name)
-            # create an app
-            cls = getattr(module, cls_name)
-            args = info.get("args", {})
-            app = cls(name, self, **args)
-            # set a slot of broadcast signal to router
-            app.broadcastRequested.connect(self._route_to_bus)
-            # add the app to the list of subscribers on each bus
-            for bus_name in info["bus"]:
-                self._subscribers[bus_name].append(app)
-            # store the app
-            self._apps[name] = app
-
-    def _show_frame(self, setup_app: dict):
-        """Shows frames of each app.
+        # create a bus
+        if "timeout" in info:
+            bus = Bus(name, info["timeout"])
+        else:
+            bus = Bus(name)
+        # set a slot of received signal to router
+        bus.received.connect(self._routeToApp)
+        bus.start()
+        # store the bus
+        self._buses[name] = bus
+        self._subscribers.setdefault(name, set())

```

Comment:
I think [`setdefault()`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) does not fit here that well.
It is used when one wants to get the value corresponding to a key, i.e. `name` here, and if the key is not in it, then add the key with the default value.
So it is similar to `get()` with default, but `get()` does not add the default value into the dict.

In this case, there must not exist `name` as I said in another comment. Therefore, using `update()` or `self._subscribers[name] = set()` would be better.

Reply:
BTW it is a good choice to use `set` 👍

Reply:
Then, how do you think about the below situation?

1. Using `load()`, initialize `Swift`.
2. Using `destroyBus()`, destroy a bus.
3. Using `createBus()`, create a bus which I removed step 2.

According to your suggestion, the subscribers info would be removed..

Reply:
Oh, I thought the subscriptions expire when a bus is destroyed.

Reply:
Ah, then what about your idea? I think the subscriptions must not expire..

Reply:
Hmm, okay then for now, let's keep your decision, since there is no way to re-subscribe the buses at the moment.
This should be discussed further in another discussion.